sdk/core/azure-json/src/main/java/com/azure/json/JsonObject.java

```diff
@@ -0,0 +1,503 @@
+                        }
+                        // Case: i is a JsonNumber object, therefore the element
+                        // is added as a JsonNumber object
+                        else if(i instanceof JsonNumber) {
+                            ((JsonArray)value).addElement((JsonNumber)i);
+                        }
+                        // Case: i is a JsonObject object, therefore the element
+                        // is added as a JsonObject object
+                        else if(i instanceof JsonObject) {
+                            ((JsonArray)value).addElement((JsonObject)i);
+                        }
+                        // Case: i is a JsonString object, therefore the element
+                        // is added as a JsonString object
+                        else if(i instanceof JsonString) {
+                            ((JsonArray)value).addElement((JsonString)i);
+                        }
+                        // Case: i is not of a type that can be casted or
+                        // converted to a valid JsonElement type. The valid
+                        // JsonElement types are:
+                        //      - JsonArray
+                        //      - JsonBoolean
+                        //      - JsonNull
+                        //      - JsonNumber
+                        //      - JsonObject
+                        //      - JsonString
+                        else { throw new IllegalArgumentException(); }
+                    }
+                }
+                // Case: element is not an array of a type that can be casted or
+                // converted to a valid JsonElement type.
+                else { throw new IllegalArgumentException(); }
+            }
+            // Case: element is not null, but is also not a valid type that can
+            // be cast or converted to a valid JsonElement type.
+            else { throw new IllegalArgumentException(); }
+        }
+        // Case: element is null, therefore value is a JsonNull object.
+        else { value = new JsonNull(); }
+
+        // Adding the new property, the key with its value pair. The value is
+        // the respective JsonElement cast/conversion of element.
+        this.addProperty(key, value);
+        //properties.put(key, value);
+
+        // Returning the new state of the JsonObject after the successful
+        // addition of the new property
+        return this;
+    }
+     */
+
+    /**
+     * Adds a new property into the JsonObject object using JsonElement as the
+     * property value.
+     *
+     * The primary user is the build method of JsonObject and JsonArray.
+     *
+     * @param key
+     * @param element
+     * @return The new state of the JsonObject after the successful addition of
+     * the new property
+     * @throws IllegalArgumentException
+     */
+    public JsonObject addProperty(String key, JsonElement element) throws IllegalArgumentException {
```

Comment:
We `return this` to allow for method chaining. I would suspect that the value prop of chaining might suggest we keep it as-is, instead of returning the `JsonElement`? Keen to discuss this before any changes are made.

Reply:
I think for methods where we are adding or setting, I don't have strong conviction on whether chaining or the overridden value should be returned. For methods where we are removing values I do think we should return the removed value as that workflow is different and the removed value may want to be used. For example, I know we use the remove APIs in Jackson and work with the removed element: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/implementation/jackson/FlatteningDeserializer.java#L125

Reply:
@Robsimpson adjust return behaviour so that removed values are returned (rather than this)

Reply:
@alzimmermsft done